### PR TITLE
Introduce a few new mimetypes for code, fix recursive mimetype aliases

### LINKED ISSUE #19254: Code files are not detected as text files with code icon.
### Steps to reproduce

As admin user do:
1. Upload code files.
2. See their icons.

### Expected behaviour
They have a code icon. Like .pl files in this screenshot:

screen shot 2015-09-21 at 13 08 26

### Actual behaviour
They are not detected as code.

.cpp, .xml and .tex are detected as plain text not code.

screen shot 2015-09-22 at 08 56 11

.abap .ads .agi .algol .e .apl .scpt .arc. ArnoldC .as .au3 .awk .bas .bat .befunge .bg .boo .bf .ceylon .clj .cbl .coffee .cfml .lisp .context **.cs** .csh .d **.dart** .pas .djs .ecl .e .ex .lisp .erl .fal .f .fy .fan .fimpp .fs **.go** .golo .gsx .groovy .hs .hs .ho .hq9plus .idr .io **.java** .jl .kt .ls .lgt .lol .lua .m4 .mathematica .m .mod .nasm .n .ml .omgrofl .pn .ps1 .pde .pure .r .rkt .rexx .rb .sml .sbl **.sql** .stata .swift **.tcl** .uc .vala .v .vim .vb .asm .xpath2 .xquery .xslt .xtend .zu

All this file types are code and are showed as undetected type, like .rtf which is not detected as a document.
I think bolded ones needs to be fixed specially since those filetypes are very used.

screen shot 2015-09-22 at 08 55 51

### Server configuration
**Operating system**:
Ubuntu 14.04

**Web server:**
Apache

**Database:**
MySQL

**PHP version:**
5.5.9

**ownCloud version:** (see ownCloud admin page)
{"installed":true,"maintenance":false,"version":"8.2.0.6","versionstring":"8.2 beta1","edition":"Enterprise"}

**Updated from an older ownCloud or fresh install:**
Fresh

**List of activated apps:**
Default in community edition..

**The content of config/config.php:**

```

```

**Are you using external storage, if yes which one:** local/smb/sftp/...
No

**Are you using encryption:** 
No

**Logs**
```
```
### Client configuration

**browser**
Chrome. Firefox.

cc @jancborchardt

### PR DESCRIPTION
**Pull request by RobinMcCorkell**
Some new/fixed mappings (and corresponding aliases):
- Java `.class`
- C++ source `.c++`
- C++ header `.hpp`
- Java source `.java`
- RTF `.rtf`

In addition, recursive mimetype aliases for getting a filetype icon are implemented, which fixes the icon for XML. application/xml -> text/html -> text/code.

cc @jancborchardt @rullzer @SergioBertolinSG 

Fixes #19254

### CODE REVIEW

**Comment by karlitschek** [comment]
I would like to question why we need this mime-types in the first place. It we don't show special icons in the web ui then this shoudn't matter.
Or do I miss something?
:-1:

**Comment by Xenopathic** [comment]
@karlitschek At the moment, some files that are code files (e.g. `.cpp`, `.hpp`, `.java`, `.xml`) are shown as plain text

File: core/js/mimetypelist.js [hunk 1]
```diff
@@ -52,56 +49,52 @@ OC.MimeTypeList={
     "application/vnd.openxmlformats-officedocument.wordprocessingml.document": "x-office/document",
     "application/vnd.openxmlformats-officedocument.wordprocessingml.template": "x-office/document",
     "application/x-7z-compressed": "package/x-generic",
-    "application/x-cbr": "text",
     "application/x-compressed": "package/x-generic",
     "application/x-dcraw": "image",
     "application/x-deb": "package/x-generic",
-    "application/x-font": "image",
+    "application/x-font": "font",
```
Comment (jancborchardt) [inline core/js/mimetypelist.js]: We do not have »font« anymore. Please use »image« instead.

Comment (jancborchardt) [inline core/js/mimetypelist.js]: Above as well.

File: core/js/mimetypelist.js [hunk 2]
```diff
@@ -52,56 +49,52 @@ OC.MimeTypeList={
...
     "application/x-tar": "package/x-generic",
     "application/x-tex": "text",
     "application/xml": "text/html",
     "application/yaml": "text/code",
     "application/zip": "package/x-generic",
-    "database": "file",
     "httpd/unix-directory": "dir",
-    "image/svg+xml": "image",
-    "image/vector": "image",
+    "image/svg+xml": "image/vector",
```
Comment (jancborchardt) [inline core/js/mimetypelist.js]: We don’t have a differentiation between vector either. Please use just »image«.

File: config/mimetypemapping.dist.json [hunk 3]
```diff
@@ -130,6 +133,7 @@
 	"raf": ["image/x-dcraw"],
 	"rar": ["application/x-rar-compressed"],
 	"reveal": ["text/reveal"],
+	"rtf": ["text/rtf"],
```
Comment (jancborchardt) [inline config/mimetypemapping.dist.json]: This is correctly mapped to use the »text« type, right? (Grey document with white lines.)

File: core/js/mimetypelist.js [hunk 4]
```diff
@@ -9,26 +9,23 @@
...
-    "application/font-sfnt": "image",
-    "application/font-woff": "image",
-    "application/illustrator": "image",
-	"application/javascript": "text/code",
+    "application/font-sfnt": "font",
+    "application/font-woff": "font",
+    "application/illustrator": "image/vector",
     "application/json": "text/code",
-    "application/msaccess": "file",
+    "application/msaccess": "database",
```
Comment (jancborchardt) [inline core/js/mimetypelist.js]: We don’t have database anymore either – use »code« instead.

**Comment by jancborchardt** [comment]
@Xenopathic in the PR for the new mimetype icons I directly edited `core/js/mimetypelist.js` https://github.com/owncloud/core/pull/18608 – it seems some of the changes are overwritten by this. For example we don’t differentiate that much anymore (no specific icons for database, vector, font etc) Can you make sure to move the changes into the source file so it’s fixed? Thanks!

File: core/js/mimetypelist.js (hunk 1 above)
Comment (Xenopathic) [inline core/js/mimetypelist.js]: I didn't touch this file manually, that was update-js doing that. I wonder what the real issue is...

File: config/mimetypemapping.dist.json (hunk 3 above)
Comment (Xenopathic) [inline config/mimetypemapping.dist.json]: Yes, if a `text-rdf` icon exists it uses that, otherwise `text`.

**Comment by Xenopathic** [comment]
@jancborchardt Ahh, so your changes were made directly to the JS file instead of the source JSON files?

**Comment by Xenopathic** [comment]
@jancborchardt Changes made :)

**Comment by jancborchardt** [comment]
> @jancborchardt Ahh, so your changes were made directly to the JS file instead of the source JSON files?

@Xenopathic yes, I am an idiot and didn’t know we have that new source JSON file. ;)

**Comment by jancborchardt** [comment]
Looks good :+1: 

@SergioBertolinSG please check if this fixes your issue.

**Comment by karlitschek** [comment]
ok. It makes sense to show the code icon. :+1:

**Comment by SergioBertolinSG** [comment]
I'm not seeing much improvement here.

.class are not text, so it can stay with the undefined icon.
.java and .rtf are now plain text which is not fine, .java is code and .rtf is document (not sure if the documents app can open it, but it should have same icon as .doc).
:-1: 

.hpp and .c++ are correct.

Many filetypes pending.

**Comment by rullzer** [comment]
Updated the c++ and cpp mimetypes. They now both map to code.
.class now maps to application/java-vm which maps to the application icon. (I stole that from http://www.freeformatter.com/mime-types-list.html)

@SergioBertolinSG you should run `./occ maintenance:mimetype:update-db` to make sure your database is also has all the updated mimetypes.

@Xenopathic feel free to squash our commits.

**Comment by Xenopathic** [comment]
@SergioBertolinSG I agree, `.class` aren't text, which is why they appear with the binary icon? `.java` is definitely a code type, not plain text. And I was asked to make `.rtf` text...

@rullzer Please note that `text/x-c++src` is not a valid mimetype as far as I can tell, there are a few references to it on the Internet but most places say to use `text/x-c`, as does the `file` utility. As for `.class`, the majority of sources used `application/java`, not `application/java-vm`. In fact, I could only find one source that said that: http://www.freeformatter.com/mime-types-list.html

**Comment by SergioBertolinSG** [comment]
@rullzer thanks, now I can see fine also .xml, .java and .cpp.

It ok for me to merge this :+1: but there are still lots pending. (some pretty popular like .cs or .rb)

Also rtf as text.

For me it doesn't fix the issue totally, but partially, yes.

**Comment by rullzer** [comment]
@Xenopathic ah ok fine by me. I just saw the discrepency between c++ and cpp. I'm fine either way. 

Regarding .class I also don't really care. Was just my first hit on google :P

**Comment by Xenopathic** [comment]
@rullzer RE cpp: oops, I was meant to change both of them then. Still, as long as the icons work properly

**Comment by Xenopathic** [comment]
@SergioBertolinSG PRs welcome :smile:

**Comment by rullzer** [comment]
@Xenopathic you are right. cpp and c++ now both map to `text/x-c` and class maps to `application/java`.

Rebased since #18893 was merged.

File: resources/config/mimetypemapping.dist.json [hunk 5]
```diff
@@ -27,15 +27,16 @@
 	"cbz": ["application/x-cbr"],
 	"cc": ["text/x-c"],
 	"cdr": ["application/coreldraw"],
+	"class": ["application/java"],
 	"cnf": ["text/plain"],
 	"conf": ["text/plain"],
-	"cpp": ["text/x-c++src"],
```
Comment (DeepDiver1975) [inline resources/config/mimetypemapping.dist.json]: cpp is c plus plus - why is this changed

File: core/js/mimetypelist.js [hunk 6]
```diff
@@ -49,10 +52,11 @@ OC.MimeTypeList={
     "application/vnd.openxmlformats-officedocument.wordprocessingml.document": "x-office/document",
     "application/vnd.openxmlformats-officedocument.wordprocessingml.template": "x-office/document",
     "application/x-7z-compressed": "package/x-generic",
+    "application/x-cbr": "text",
     "application/x-compressed": "package/x-generic",
     "application/x-dcraw": "image",
     "application/x-deb": "package/x-generic",
-    "application/x-font": "font",
```
Comment (DeepDiver1975) [inline core/js/mimetypelist.js]: font is font - what is the problem with that?

Comment (Xenopathic) [inline core/js/mimetypelist.js]: Apparently we don't use `font` anymore, just `image` https://github.com/owncloud/core/pull/19272#discussion_r40088783 :see_no_evil:

File: resources/config/mimetypemapping.dist.json (hunk 5 above)
Comment (Xenopathic) [inline resources/config/mimetypemapping.dist.json]: The generally accepted mimetype for c++ files is text/x-c

Comment (rullzer) [inline resources/config/mimetypemapping.dist.json]: warg depending on the source you check it is text/x-c++src 
and even text/x-csrc for c files...

**Comment by rullzer** [comment]
Ok... so which list of mimetypes do we actually follow?

**Comment by Xenopathic** [comment]
@rullzer In case there's conflicts, I look at what `file -i` prints (which is why I chose `text/x-c` over `text/x-c++src`)

**Comment by rullzer** [comment]
> @rullzer In case there's conflicts between sources on the web, I look at what file -i prints (which is why I chose text/x-c over text/x-c++src)

@Xenopathic but when I issue the `mimetype` command on `foo.c++` I do get `text/x-c++src` here...

**Comment by Xenopathic** [comment]
@rullzer That's strange, I get `text/x-c` for both `.c++` and `.cpp`...

**Comment by Xenopathic** [comment]
Wait, `mimetype` command? What's that?

**Comment by rullzer** [comment]
@Xenopathic ah I have that on my system. Basically what is does is look trough `/usr/share/mime`

**Comment by Xenopathic** [comment]
@rullzer Ah, that's where mimetypes are stored. I see the `text/x-c++src` mimetype now. In that case, we should use that (it's strange that `file -i` doesn't use /usr/share/mime though...)

**Comment by Xenopathic** [comment]
Also, being radical for a moment: could ownCloud directly use /usr/share/mime?

**Comment by rullzer** [comment]
@Xenopathic no. because it is not quaranteed to exist. So probabaly `file -i` retrieves it mimetypes from somewhere else. Basically `/usr/share/mime` is part of the freedesktop stuff.

But we could maybe write a script that just parses the stuff from https://wiki.freedesktop.org/www/Software/shared-mime-info/ and creates the proper mappings.

**Comment by rullzer** [comment]
Anyway. Back on topic. `c++` and `cpp` now map to `text/x-c++src` again.

**Comment by Xenopathic** [comment]
@jancborchardt @SergioBertolinSG Is this complete enough for merge? I know it doesn't cover the full list of mimetypes, but it covers the important ones.

**Comment by rullzer** [comment]
Rebases to make sure there are no conflicts.
Lets get this in. If more mappings are needed eventually we can add those later.

Easy to review!

CC: @PVince81 @MorrisJobke @nickvergessen

File: resources/config/mimetypemapping.dist.json [hunk 7]
```diff
@@ -130,6 +133,8 @@
 	"raf": ["image/x-dcraw"],
 	"rar": ["application/x-rar-compressed"],
 	"reveal": ["text/reveal"],
+	"rss": ["application/rss+xml"],
+	"rtf": ["text/rtf"],
```
Comment (MorrisJobke) [inline resources/config/mimetypemapping.dist.json]: I think we need repair steps for the added extensions. Add it with a version number check in https://github.com/owncloud/core/blob/83db69b45d3ff5f49634839c0deafde91872a6b5/lib/repair/repairmimetypes.php#L264-264

But please raise the version number and check for everything that is smaller than the new one.

Comment (rullzer) [inline resources/config/mimetypemapping.dist.json]: You are right. Will fix.

**Comment by rullzer** [comment]
@MorrisJobke added.

**Comment by MorrisJobke** [comment]
Tested and works :+1:

**Comment by MorrisJobke** [comment]
@SergioBertolinSG Please review :)

**Comment by icewind1991** [comment]
:+1:

**Comment by SergioBertolinSG** [comment]
Well I only see .java and .cpp files correct. Tex files has a text icon but cannot be opened. Many more remaining (.cs, .sql, .go...).

**Comment by SergioBertolinSG** [comment]
Issue for tracking pending mimetypes: https://github.com/owncloud/core/issues/19573

**Pull request merged**